File: java/3. Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> hset = new HashSet();
        int windowStart = 0, maxLength = 0;
        for(int windowEnd = 0; windowEnd < s.length(); windowEnd++) {
            while(hset.contains(s.charAt(windowEnd))) {
                hset.remove(s.charAt(windowStart));
                windowStart++;
            }
            hset.add(s.charAt(windowEnd));
            maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
        }
        return maxLength; 
    }
}

File: python/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        h_set = set()
        window_start = 0
        max_len = 0
        for window_end in range(len(s)):
            while s[window_end] in h_set:
                h_set.remove(s[window_start])
                window_start += 1
            h_set.add(s[window_end])
            max_len = max(window_end - window_start + 1, max_len)
        return max_len
    

def main():
        solution = Solution()
        #TC1 - 3
        print(solution.lengthOfLongestSubstring("abcabcbb"))
        #TC2 - 1
        print(solution.lengthOfLongestSubstring("bbbbb"))
        #TC3 - 3
        print(solution.lengthOfLongestSubstring("pwwkew"))


if __name__ == "__main__":
    main()